nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/actions/SetPropertiesAction.java

```diff
@@ -76,92 +64,41 @@
-        }
-
-        @Override
-        public void processRecord(ComputationContext context, String inputStreamName, Record record) {
-            String commandId = commandIdFrom(record);
-            if (currentCommandId == null) {
-                // first time we need to process something
-                loadCurrentBulkCommandContext(commandId);
-            } else if (!currentCommandId.equals(commandId)) {
-                // new bulk id computation - send remaining elements
-                processBatch(context);
-                documentIds.clear();
-                loadCurrentBulkCommandContext(commandId);
-            }
-            // process record
-            documentIds.addAll(docIdsFrom(record));
-            if (documentIds.size() >= batchSize) {
-                processBatch(context);
-            }
-        }
-
-        protected void loadCurrentBulkCommandContext(String commandId) {
-            currentCommandId = commandId;
-            KeyValueStore kvStore = Framework.getService(KeyValueService.class).getKeyValueStore(BULK_KV_STORE_NAME);
-            currentCommand = BulkCodecs.getBulkCommandCodec().decode(kvStore.get(commandId + COMMAND));
-        }
-
-        @Override
-        public void destroy() {
-            log.debug(String.format("Destroy computation: %s, pending entries: %d", SETPROPERTIES_ACTION_NAME,
-                    documentIds.size()));
-        }
-
         protected void processBatch(ComputationContext context) {
             if (!documentIds.isEmpty()) {
                 TransactionHelper.runInTransaction(() -> {
-                    // for setProperties, parameters are properties to set
-                    Map<String, Serializable> properties = currentCommand.getParams();
-                    LoginContext loginContext;
                     try {
-                        loginContext = Framework.loginAsUser(currentCommand.getUsername());
-
-                        try (CloseableCoreSession session = CoreInstance.openCoreSession(
-                                currentCommand.getRepository())) {
-                            for (String docId : documentIds) {
-                                DocumentModel doc = session.getDocument(new IdRef(docId));
-                                properties.forEach(doc::setPropertyValue);
-                                session.saveDocument(doc);
-                            }
+                        LoginContext loginContext = Framework.loginAsUser(currentCommand.getUsername());
+                        String repository = currentCommand.getRepository();
+                        // for setProperties, parameters are properties to set
+                        Map<String, Serializable> properties = currentCommand.getParams();
+                        try (CloseableCoreSession session = CoreInstance.openCoreSession(repository)) {
+                            doProcessBatch(session, documentIds, properties);
```

Comment:
You can probably move things about transaction, login and session to `AbstractBulkComputation`.

Reply:
To discuss, probably same move to do for counter as we leverage it to detect end.

Reply:
In principle I'm all for it, and I thought about moving the whole processBatch into the AbstractBulkComputation and let only doProcessBatch in SetProperties but wasn't sure it would fit next implementations and wanted to give us time to improve the abstraction while doing them.
wdyt ?